File: client-samples/web/StreamKit/Backends/LiveKit/LiveKitBackend.js
/**
 * @fileoverview LiveKit WebRTC backend for StreamKit.
 *
 * Implements the {@link StreamingBackend} interface using the LiveKit JS SDK v2.
 * The SDK is resolved via an import map in the host page's `index.html`; no
 * build step is required.
 *
 * Mirror of the Swift `LiveKitBackend` class. Three deliberate platform
 * adaptations apply — see `.claudelearnings/decisions/web-streamkit-api-design.md`.
 *
 * @module StreamKit/Backends/LiveKit/LiveKitBackend
 */

import {
  Room,
  RoomEvent,
  Track,
  createLocalVideoTrack,
  createLocalAudioTrack,
} from 'livekit-client';

import { ConnectionState } from '../../ConnectionState.js';
import { StreamError } from '../../StreamError.js';
import { MicrophoneMode } from '../../Config/AudioConfig.js';

// ─────────────────────────────────────────────────────────────────────────────

/**
 * Maps a LiveKit `ConnectionState` string to the StreamKit {@link ConnectionState}
 * enum value.
 *
 * LiveKit v2 fires `RoomEvent.ConnectionStateChanged` with one of the literal
 * strings `'connected'`, `'connecting'`, `'disconnected'`, or `'reconnecting'`.
 * Those happen to match our own enum values one-for-one, but we map explicitly
 * here rather than relying on that coincidence, so a future SDK bump won't
 * silently produce wrong states.
 *
 * @param {string} lkState - Raw LiveKit ConnectionState string.
 * @returns {string} A {@link ConnectionState} value.
 */
function mapState(lkState) {
  switch (lkState) {
    case 'connected':     return ConnectionState.CONNECTED;
    case 'connecting':    return ConnectionState.CONNECTING;
    case 'reconnecting':  return ConnectionState.RECONNECTING;
    case 'disconnected':
    default:              return ConnectionState.DISCONNECTED;
  }
}

// ─────────────────────────────────────────────────────────────────────────────

/**
 * Implements {@link StreamingBackend} over LiveKit WebRTC.
 *
 * @implements {import('../StreamingBackend.js').StreamingBackend}
 *
 * @example
 * // Consumers should not construct this directly — use BackendConfiguration:
 * import { BackendConfiguration, LiveKitConfig } from '../../Config/BackendConfiguration.js';
 *
 * const session = await StreamSession.create(
 *   BackendConfiguration.liveKit(new LiveKitConfig({ host: '192.168.1.100', token: jwt }))
 * );
 */
export class LiveKitBackend {
  // ── Private fields ──────────────────────────────────────────────────────────

  /** @type {import('../../Config/BackendConfiguration.js').LiveKitConfig} */
  #config;

  /** @type {Room | null} */
  #room = null;

  /** @type {import('livekit-client').LocalVideoTrack | null} */
  #videoTrack = null;

  /** @type {import('livekit-client').LocalAudioTrack | null} */
  #audioTrack = null;

  /** @type {import('../../Config/SessionConfig.js').SessionConfig | null} */
  #sessionConfig = null;

  // ── Public event hooks ──────────────────────────────────────────────────────

  /**
   * Called whenever the connection lifecycle state changes.
   * `StreamSession` assigns this before calling `connect()`.
   *
   * @type {((state: string) => void) | null}
   */
  onConnectionStateChanged = null;

  /**
   * Called whenever binary data is received from remote participants.
   * `StreamSession` assigns this before calling `connect()`.
   *
   * @type {((data: Uint8Array) => void) | null}
   */
  onDataReceived = null;

  // ── Constructor ─────────────────────────────────────────────────────────────

  /**
   * @param {import('../../Config/BackendConfiguration.js').LiveKitConfig} config
   */
  constructor(config) {
    this.#config = config;
  }

  // ── Public API ──────────────────────────────────────────────────────────────

  /**
   * Establishes a LiveKit room connection.
   *
   * Tears down any existing session first, validates the host, builds the
   * WebSocket URL, acquires a JWT (from `config.token` or the token endpoint),
   * creates and wires up a `Room`, then connects. Audio is published
   * automatically unless the session config disables the microphone.
   *
   * @param {import('../../Config/SessionConfig.js').SessionConfig} sessionConfig
   * @returns {Promise<void>}
   * @throws {StreamError} `invalidHost` | `missingToken` | `tokenFetchFailed`
   */
  async connect(sessionConfig) {
    // Tear down any existing session cleanly before starting a new one.
    await this.#tearDown();

    this.#sessionConfig = sessionConfig;

    // ── Validate host & build WebSocket URL ───────────────────────────────────
    const { host, port, secure } = this.#config;
    if (!host || typeof host !== 'string' || host.trim() === '') {
      throw StreamError.invalidHost(host);
    }

    // Reject strings that already contain a scheme or look structurally wrong.
    let cleanHost;
    try {
      // Temporarily add a scheme so URL can parse the authority part.
      const probe = new URL(`ws://${host}`);
      cleanHost = probe.hostname;
      if (!cleanHost) throw new Error('empty hostname');
    } catch {
      throw StreamError.invalidHost(host);
    }

    const scheme = secure ? 'wss' : 'ws';
    const wsURL = `${scheme}://${cleanHost}:${port}`;

    // ── Resolve JWT ───────────────────────────────────────────────────────────
    let token;
    if (this.#config.token) {
      token = this.#config.token;
    } else if (this.#config.tokenURL) {
      token = await this.#fetchToken(this.#config.tokenURL, sessionConfig.identity);
    } else {
      throw StreamError.missingToken();
    }

    // ── Create Room and wire events ───────────────────────────────────────────
    const room = new Room();
    this.#room = room;

    room.on(RoomEvent.ConnectionStateChanged, (lkState) => {
      this.onConnectionStateChanged?.(mapState(lkState));
    });

    room.on(RoomEvent.DataReceived, (payload /*, participant, kind, topic */) => {
      // payload is already a Uint8Array per the LiveKit v2 API.
      this.onDataReceived?.(payload);
    });

    // ── Connect ───────────────────────────────────────────────────────────────
    await room.connect(wsURL, token);

    // ── Publish audio (unless disabled) ──────────────────────────────────────
    const audioMode = sessionConfig.audio?.mode;
    if (audioMode !== MicrophoneMode.DISABLED) {
      await this.#publishAudio(sessionConfig);
    }
  }

  /**
   * Disconnects from the room and releases all resources.
   *
   * Safe to call at any time, including before `connect()`.
   *
   * @returns {Promise<void>}
   */
  async disconnect() {
    await this.#tearDown();
  }

  /**
   * Captures the local camera and publishes it to the room.
   *
   * The `facingMode` is taken from `sessionConfig.camera.facing`, which is
   * already a browser `getUserMedia` constraint value (`'user'`/`'environment'`).
   *
   * @returns {Promise<void>}
   * @throws {StreamError} `cameraRequiresConnection` if not connected.
   */
  async startCamera() {
    if (!this.#room || this.#room.state !== 'connected') {
      throw StreamError.cameraRequiresConnection();
    }

    // Stop any previous video track first.
    await this.stopCamera();

    const facingMode = this.#sessionConfig?.camera?.facing ?? 'user';
    const track = await createLocalVideoTrack({ facingMode });
    this.#videoTrack = track;

    await this.#room.localParticipant.publishTrack(track, {
      source: Track.Source.Camera,
    });
  }

  /**
   * Unpublishes and stops the local video track.
   *
   * Resolves silently if the camera was never started.
   *
   * @returns {Promise<void>}
   */
  async stopCamera() {
    if (!this.#videoTrack) return;

    if (this.#room) {
      await this.#room.localParticipant.unpublishTrack(this.#videoTrack);
    }
    this.#videoTrack.stop();
    this.#videoTrack = null;
  }

  /**
   * Sends binary data to remote participants via the LiveKit data channel.
   *
   * Accepts a `string`, `ArrayBuffer`, or `Uint8Array`. Strings are encoded
   * to UTF-8. Keep individual messages under 15 KB (LiveKit's WebRTC MTU)
   * to avoid fragmentation.
   *
   * @param {string | ArrayBuffer | Uint8Array} data
   * @param {object}  [opts]
   * @param {boolean} [opts.reliable=true]
   * @returns {Promise<void>}
   * @throws {StreamError} `notConnected` if not connected.
   */
  async send(data, { reliable = true } = {}) {
    if (!this.#room || this.#room.state !== 'connected') {
      throw StreamError.notConnected();
    }

    let bytes;
    if (typeof data === 'string') {
      bytes = new TextEncoder().encode(data);
    } else if (data instanceof Uint8Array) {
      bytes = data;
    } else {
      // ArrayBuffer or any other BufferSource.
      bytes = new Uint8Array(data);
    }

    await this.#room.localParticipant.publishData(bytes, { reliable });
  }

  // ── Private helpers ─────────────────────────────────────────────────────────

  /**
   * Disconnects the room, stops all tracks, and nulls out all state.
   *
   * @returns {Promise<void>}
   */
  async #tearDown() {
    if (this.#room) {
      // Remove all listeners before disconnecting to prevent stale callbacks.
      this.#room.removeAllListeners();
      await this.#room.disconnect();
      this.#room = null;
    }

    if (this.#videoTrack) {
      this.#videoTrack.stop();
      this.#videoTrack = null;
    }

    if (this.#audioTrack) {
      this.#audioTrack.stop();
      this.#audioTrack = null;
    }

    this.#sessionConfig = null;
  }

  /**
   * Creates and publishes a local audio track based on the session's
   * `AudioConfig`. Called automatically from `connect()`.
   *
   * @param {import('../../Config/SessionConfig.js').SessionConfig} sessionConfig
   * @returns {Promise<void>}
   */
  async #publishAudio(sessionConfig) {
    const { mode, echoCancellation } = sessionConfig.audio;

    // Derive WebRTC constraint booleans from MicrophoneMode.
    // VOICE_PROCESSING: enable the full browser voice-isolation pipeline.
    // SOFTWARE_PROCESSING: standard WebRTC DSP — use echoCancellation setting.
    // RAW: disable all DSP so the server can handle processing.
    const isVoice = mode === MicrophoneMode.VOICE_PROCESSING;
    const isSoftware = mode === MicrophoneMode.SOFTWARE_PROCESSING;

    const track = await createLocalAudioTrack({
      echoCancellation: isVoice ? true : (isSoftware ? echoCancellation : false),
      noiseSuppression: isVoice || isSoftware,
      autoGainControl:  isVoice || isSoftware,
    });

    this.#audioTrack = track;

    await this.#room.localParticipant.publishTrack(track, {
      source: Track.Source.Microphone,
    });
  }

  /**
   * Fetches a LiveKit JWT from a token endpoint.
   *
   * The endpoint is called as:
   * `GET <tokenURL>?identity=<identity>`
   *
   * The server must return either a plain JWT string or a JSON object with a
   * `"token"` key:  `{ "token": "eyJ…" }`.
   *
   * @param {string} tokenURL  - Base URL of the token endpoint.
   * @param {string} identity  - Participant identity to embed in the token.
   * @returns {Promise<string>} Resolved JWT string.
   * @throws {StreamError} `tokenFetchFailed` on any network or parse error.
   */
  async #fetchToken(tokenURL, identity) {
    let response;
    try {
      const url = new URL(tokenURL);
      url.searchParams.set('identity', identity);
      response = await fetch(url.toString());
    } catch (err) {
      throw StreamError.tokenFetchFailed(tokenURL);
    }

    if (!response.ok) {
      throw StreamError.tokenFetchFailed(tokenURL);
    }

    const text = await response.text();

    // Try JSON first: { "token": "eyJ…" }
    try {
      const json = JSON.parse(text);
      if (json && typeof json.token === 'string' && json.token.length > 0) {
        return json.token;
      }
    } catch {
      // Not JSON — fall through to plain-string check.
    }

    // Plain JWT string.
    const trimmed = text.trim();
    if (trimmed.length > 0) {
      return trimmed;
    }

    throw StreamError.tokenFetchFailed(tokenURL);
  }
}

File: client-samples/android/app/src/main/java/com/nvidia/xrai/streamkitsample/streamkit/backends/livekit/LiveKitBackend.kt
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.nvidia.xrai.streamkitsample.streamkit.backends.livekit

import android.content.Context
import com.nvidia.xrai.streamkitsample.streamkit.ConnectionState
import com.nvidia.xrai.streamkitsample.streamkit.StreamError
import com.nvidia.xrai.streamkitsample.streamkit.backends.StreamingBackend
import com.nvidia.xrai.streamkitsample.streamkit.config.AudioConfig
import com.nvidia.xrai.streamkitsample.streamkit.config.BackendConfiguration
import com.nvidia.xrai.streamkitsample.streamkit.config.CameraConfig
import com.nvidia.xrai.streamkitsample.streamkit.config.LiveKitConfig
import com.nvidia.xrai.streamkitsample.streamkit.config.SessionConfig
import io.livekit.android.LiveKit
import io.livekit.android.events.RoomEvent
import io.livekit.android.events.collect
import io.livekit.android.renderer.TextureViewRenderer
import io.livekit.android.room.Room
import io.livekit.android.room.track.CameraPosition
import io.livekit.android.room.track.DataPublishReliability
import io.livekit.android.room.track.LocalVideoTrack
import io.livekit.android.room.track.LocalVideoTrackOptions
import io.livekit.android.room.track.Track
import io.livekit.android.room.track.VideoCaptureParameter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.URL
import java.net.URLEncoder
import java.nio.ByteBuffer

/**
 * [StreamingBackend] implementation using the LiveKit Android SDK.
 *
 * Consumers should not construct this directly — create it via
 * [BackendConfiguration.LiveKit] passed to [StreamSession][com.nvidia.xrai.streamkitsample.streamkit.StreamSession].
 *
 * Mirror of the Swift `LiveKitBackend` and web `LiveKitBackend`.
 *
 * ## Remote audio
 * LiveKit Android renders remote audio tracks automatically once subscribed —
 * no explicit attachment step is required (unlike the web SDK).
 *
 * ## Connection-state model
 * LiveKit Android `Room.connect()` is a suspend function that returns when the
 * room is CONNECTED (or throws). We emit CONNECTING before calling it and
 * CONNECTED after it returns. RECONNECTING / DISCONNECTED come through
 * `room.events` asynchronously.
 */
internal class LiveKitBackend(
    private val config: LiveKitConfig,
    private val appContext: Context,
) : StreamingBackend {

    // ── Event hooks ────────────────────────────────────────────────────────────

    override var onConnectionStateChanged: ((ConnectionState) -> Unit)? = null
    override var onDataReceived: ((topic: String, data: ByteArray) -> Unit)? = null
    override var onAgentStatus: ((status: String) -> Unit)? = null

    // ── Public local preview accessor ─────────────────────────────────────────

    /**
     * Currently published local camera track, or null when stopped.  Used by
     * `CameraPreviewView` to render the outgoing stream locally; app code
     * goes through that composable rather than touching this directly.
     */
    val localCameraTrack: LocalVideoTrack?
        get() = room?.localParticipant
            ?.getTrackPublication(Track.Source.CAMERA)?.track as? LocalVideoTrack

    /** Initialises a [TextureViewRenderer] with the connected room's EGL
     *  context so it can sink frames from the local camera track. No-op when
     *  the room is not connected. */
    fun initVideoRenderer(view: TextureViewRenderer) {
        room?.initVideoRenderer(view)
    }

    // ── Private state ──────────────────────────────────────────────────────────

    private var room: Room? = null
    private var isConnected = false

    /** Coroutine scope active for the lifetime of one connection. */
    private var connectionScope: CoroutineScope? = null

    // ── Injected video state (external frame source) ───────────────────────────

    /** Custom capturer that lets us push externally-sourced frames into LiveKit. */
    @Volatile private var injectedCapturer: InjectedVideoCapturer? = null
    private var injectedVideoTrack: LocalVideoTrack? = null
    private val injectedVideoMutex = Mutex()

    // ── StreamingBackend: connect / disconnect ─────────────────────────────────

    override suspend fun connect(sessionConfig: SessionConfig) {
        tearDown()

        if (config.host.isBlank()) throw StreamError.InvalidHost(config.host)

        val wsUrl = "wss://${config.host}:${config.port}"

        val token = when {
            !config.token.isNullOrBlank() -> config.token
            !config.tokenURL.isNullOrBlank() -> fetchToken(config.tokenURL, sessionConfig.identity)
            else -> throw StreamError.MissingToken
        }

        val newRoom = LiveKit.create(appContext)
        room = newRoom

        val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
        connectionScope = scope

        // Collect all room events for state changes and data.
        // Must be launched before room.connect() so we don't miss RECONNECTING /
        // DISCONNECTED events that arrive immediately after a connect.
        scope.launch {
            newRoom.events.collect { event ->
                handleEvent(event)
            }
        }

        // Emit CONNECTING before the async handshake.
        onConnectionStateChanged?.invoke(ConnectionState.CONNECTING)

        // room.connect() suspends until the room is fully connected or throws.
        newRoom.connect(wsUrl, token)

        // Successfully connected.
        isConnected = true
        onConnectionStateChanged?.invoke(ConnectionState.CONNECTED)
    }

    override suspend fun disconnect() {
        tearDown()
    }

    // ── StreamingBackend: audio ────────────────────────────────────────────────

    override suspend fun startAudio(config: AudioConfig) {
        if (!isConnected) throw StreamError.NotConnected
        val enabled = config.mode != AudioConfig.MicrophoneMode.DISABLED
        room?.localParticipant?.setMicrophoneEnabled(enabled)
    }

    override suspend fun stopAudio() {
        room?.localParticipant?.setMicrophoneEnabled(false)
    }

    // ── StreamingBackend: camera ───────────────────────────────────────────────

    override suspend fun startCamera(config: CameraConfig) {
        if (!isConnected) throw StreamError.CameraRequiresConnection

        val lp = room?.localParticipant ?: return
        val position = when (config.facing) {
            CameraConfig.CameraFacing.FRONT -> CameraPosition.FRONT
            CameraConfig.CameraFacing.BACK  -> CameraPosition.BACK
        }
        // setCameraEnabled() reads the current videoTrackCaptureDefaults when it
        // creates the track. Apply both deviceId (if pinned) and position there
        // before flipping it on. Per LocalVideoTrackOptions docs, deviceId is
        // preferred — position is only the fallback when deviceId is null or
        // not found.
        lp.videoTrackCaptureDefaults = lp.videoTrackCaptureDefaults.copy(
            deviceId = config.deviceId,
            position = position,
        )
        lp.setCameraEnabled(true)
    }

    override suspend fun stopCamera() {
        injectedVideoMutex.withLock {
            val track = injectedVideoTrack
            if (track != null) {
                // unpublishTrack with default stopOnUnpublish=true already
                // calls track.stop() + track.dispose() — calling them again
                // races MediaCodec's event handler against its torn-down
                // thread ("Handler on a dead thread" IllegalStateException).
                room?.localParticipant?.unpublishTrack(track)
                injectedVideoTrack = null
                injectedCapturer = null
                return
            }
        }
        room?.localParticipant?.setCameraEnabled(false)
    }

    // ── StreamingBackend: injected video frames ───────────────────────────────

    override suspend fun injectVideoFrame(
        i420: ByteBuffer,
        width: Int,
        height: Int,
        timestampUs: Long,
    ) {
        if (!isConnected) throw StreamError.NotConnected

        // Fast path — track already exists. Take the same mutex that
        // stopCamera()/tearDown() use when they clear and dispose the
        // capturer, so a concurrent teardown cannot dispose the native
        // capturer between the null-check and pushI420Frame (use-after-dispose).
        injectedVideoMutex.withLock {
            injectedCapturer?.let {
                it.pushI420Frame(i420, width, height, timestampUs)
                return
            }
        }
        // Slow path — first frame: create + publish track under the mutex.
        val capturer = injectedVideoMutex.withLock {
            injectedCapturer?.let { return@withLock it }

            val lp = room?.localParticipant ?: throw StreamError.NotConnected
            val newCapturer = InjectedVideoCapturer()
            val track = lp.createVideoTrack(
                name = "injected-video",
                capturer = newCapturer,
                options = LocalVideoTrackOptions(
                    isScreencast = false,
                    captureParams = VideoCaptureParameter(width, height, 30),
                ),
            )
            track.startCapture()
            lp.publishVideoTrack(track)
            injectedVideoTrack = track
            injectedCapturer = newCapturer
            newCapturer
        }
        capturer.pushI420Frame(i420, width, height, timestampUs)
    }

    // ── StreamingBackend: data channel ─────────────────────────────────────────

    override suspend fun send(data: ByteArray, reliable: Boolean) {
        if (!isConnected) throw StreamError.NotConnected
        // DataPublishOptions carries reliability and optional topic.
        // The hub reads the data channel payload directly — no topic encoding needed.
        room?.localParticipant?.publishData(
            data,
            if (reliable) DataPublishReliability.RELIABLE else DataPublishReliability.LOSSY,
        )
    }

    // ── Event dispatcher ───────────────────────────────────────────────────────

    private fun handleEvent(event: RoomEvent) {
        when (event) {
            is RoomEvent.Reconnecting -> {
                onConnectionStateChanged?.invoke(ConnectionState.RECONNECTING)
            }
            is RoomEvent.Reconnected -> {
                isConnected = true
                onConnectionStateChanged?.invoke(ConnectionState.CONNECTED)
            }
            is RoomEvent.Disconnected -> {
                isConnected = false
                onConnectionStateChanged?.invoke(ConnectionState.DISCONNECTED)
            }
            is RoomEvent.DataReceived -> handleData(event)
            else -> {}
        }
    }

    private fun handleData(event: RoomEvent.DataReceived) {
        val topic = event.topic ?: ""
        val data  = event.data

        // Intercept the reserved agent-status topic — never forward to onDataReceived.
        if (topic == AGENT_STATUS_TOPIC) {
            try {
                val json   = JSONObject(String(data, Charsets.UTF_8))
                val status = json.optString("status")
                if (status.isNotEmpty()) onAgentStatus?.invoke(status)
            } catch (_: Exception) {
                // Malformed payload — ignore.
            }
            return
        }

        onDataReceived?.invoke(topic, data)
    }

    // ── Teardown ──────────────────────────────────────────────────────────────

    private suspend fun tearDown() {
        isConnected = false
        connectionScope?.cancel()
        connectionScope = null
        // Drop any injected video track before disconnecting the room. The
        // room teardown itself unpublishes tracks (which stops + disposes
        // them); avoid a manual stop()/dispose() here so we don't race the
        // MediaCodec event handler against its own torn-down thread.
        injectedVideoMutex.withLock {
            val track = injectedVideoTrack
            if (track != null) {
                runCatching { room?.localParticipant?.unpublishTrack(track) }
            }
            injectedVideoTrack = null
            injectedCapturer = null
        }
        room?.disconnect()
        room = null
        // Emit DISCONNECTED only when we initiated the teardown — a network-driven
        // disconnect will have already fired via RoomEvent.Disconnected.
        onConnectionStateChanged?.invoke(ConnectionState.DISCONNECTED)
    }

    // ── Token fetch ───────────────────────────────────────────────────────────

    /**
     * Fetches a LiveKit JWT from a token endpoint.
     *
     * Called as `GET <tokenURL>?identity=<identity>`. The endpoint must return
     * either a plain JWT string or `{ "token": "eyJ…" }`.
     *
     * Mirrors the web [fetchToken] and Swift [fetchToken] implementations exactly.
     */
    private suspend fun fetchToken(tokenURL: String, identity: String): String =
        withContext(Dispatchers.IO) {
            val encodedIdentity = URLEncoder.encode(identity, "UTF-8")
            val separator       = if (tokenURL.contains('?')) '&' else '?'
            val urlStr          = "$tokenURL${separator}identity=$encodedIdentity"

            val conn = URL(urlStr).openConnection().apply {
                connectTimeout = 5_000
                readTimeout    = 5_000
                (this as? java.net.HttpURLConnection)?.requestMethod = "GET"
            } as java.net.HttpURLConnection

            try {
                val code = conn.responseCode
                val stream = if (code in 200..299) conn.inputStream else conn.errorStream
                val body = stream?.bufferedReader()?.use { it.readText() } ?: ""

                if (code !in 200..299) {
                    throw StreamError.TokenFetchFailed(
                        urlStr,
                        "HTTP $code${if (body.isNotBlank()) ": ${body.take(200)}" else ""}",
                    )
                }

                // JSON form: { "token": "eyJ…" }
                try {
                    val t = JSONObject(body).optString("token")
                    if (t.isNotEmpty()) return@withContext t
                } catch (_: Exception) { /* not JSON, fall through */ }

                // Plain JWT form. Reject HTML/JSON-shaped bodies that didn't have a token.
                val trimmed = body.trim()
                if (trimmed.isNotEmpty() && !trimmed.startsWith("{") && !trimmed.startsWith("<")) {
                    return@withContext trimmed
                }

                throw StreamError.TokenFetchFailed(
                    urlStr,
                    "Unexpected response body: ${body.take(200)}",
                )
            } catch (e: StreamError) {
                throw e
            } catch (e: Exception) {
                // Surface the underlying failure (DNS, ConnectException, SSL, …) — the
                // generic "Token fetch failed" was useless for debugging.
                throw StreamError.TokenFetchFailed(
                    urlStr,
                    "${e.javaClass.simpleName}: ${e.message ?: "no message"}",
                )
            } finally {
                conn.disconnect()
            }
        }

    // ── Constants ──────────────────────────────────────────────────────────────

    companion object {
        /** Reserved LiveKit topic for internal agent status messages. Matches web and Swift. */
        private const val AGENT_STATUS_TOPIC = "_agent.status"
    }
}
